### ISSUE TITLE
Handling VM restarts

### ISSUE THREAD

**Issue by jellonek** [Mirantis/virtlet#334]
We have (at last) two issues connected with VM restarting. One is connected with killing `vmwrapper`, second occurs when VM halts intentionally, without action taken by kubernetes scheduler.

So effect of first issue can be seen on screen attached to #285. Once pod was started, network configuration was prepared by cni plugin, then it was disassembled before starting VM, and should be reassembled again, after VM dies. If wrapper was killed - there was no reassembling of routes, so on next query from scheduler to start container again in the same pod sandbox - network configuration is a bit messy. I think that this situation should be detected by `vmwrapper` and if there is already tap device - stripping original network config should be skipped. Also if we would want to remove pod when vmwrapper was killed - there will be no configuration on veth devices, so CNI tear down will also fail.

Second issue occurs when `qemu` will finish because of VM (guest os halts, boot fails, similar cases without infinite loop at the end). In this situation - vmwrapper "does the job" with networking, scheduler asks for restart of container in pod sandbox - virtlet tries to create new root volume and it fails, because there is one with same name, in not yet garbage collected old domain. This looks like we need modify `CreateContainer` which should test if there was already defined domain and if so - a) if it's started (probably it should not be) - it needs to be killed? b) if it's stopped - should we recreate root volume from base image, or should we try to simply restart the domain? (imo we should recreate root volume, keeping other things like network configuration not touched).

**Comment by vefimova** [comment]
re #2 issue:
I haven't had time to investigate more deeply, but after last discussion with Ivan re cleanups during Creating and Starting domains, had digged into this part of code in kuberuntime and dockershim, and noticed that actually we don't follow from the start the common logic, i.e. we don't use the number of attempts in domain's names and related to them artifacts (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/naming.go#L69), that can cause kubelet's GC doesn't work as designed and as it works for docker pods:

```root@kube-node-1:/# docker ps | grep nginx
d01f286b0cbf        nginx@sha256:41ad9967ea448d7c2b203c699b429abe1ed5af331cd92533900c6d77490e0268    "nginx -g 'daemon off"   57 seconds ago      Up 56 seconds   k8s_web_image-server-1782580915-mh7sb_kube-system_c7c5eb16-54ee-11e7-8684-02420ac00002_3
 
root@kube-node-1:/# docker stop d01f286b0cbf
d01f286b0cbf
root@kube-node-1:/# docker ps -a | grep nginx
ff01f06f374c        nginx@sha256:41ad9967ea448d7c2b203c699b429abe1ed5af331cd92533900c6d77490e0268   "nginx -g 'daemon off"   35 seconds ago      Up 34 seconds      k8s_web_image-server-1782580915-mh7sb_kube-system_c7c5eb16-54ee-11e7-8684-02420ac00002_4
d01f286b0cbf        nginx@sha256:41ad9967ea448d7c2b203c699b429abe1ed5af331cd92533900c6d77490e0268   "nginx -g 'daemon off"   2 minutes ago       Exited (0) About a minute ago   k8s_web_image-server-1782580915-mh7sb_kube-system_c7c5eb16-54ee-11e7-8684-02420ac00002_3
root@kube-node-1:/# 
```

**Comment by jellonek** [comment]
IMO second issue is not connected with GC/naming (but that would be nice to use same mechanism on our side so it deserves for separate issue).

**Comment by ivan4th** [comment]
1 - what happens if vmwrapper gets killed while the network configuration is not complete? Maybe we can think about trying to restore the initial state first? We can for sure create an issue for it for later and go with approach you suggest though, as this is probably a rare occasion.

2 - maybe we should also move domain setup logic to `StartContainer()`? It currently tries to fix the problem with existing domain by removing the domain, but I feel that could not be a right approach. I think this logic should detect non-clean container start (some leftover stuff) and remove it if it exists. It should probably look at container attempt number, yes, as upon the initial attempt the state will be clean.

**Comment by vefimova** [comment]
well, it is, restarting container with different name as well as id, would fix the second issue, if CG will clean previous container by itself.

**Comment by jellonek** [comment]
hmmm... point taken, starting new domain with new id (and new root volume) will solve most of second problem - but what with other volumes, especially with persistent ones? they probably also need to be disconnected from previously running domain (i'm not sure, but probably that would require dropping this domain), so they could be plugged into new one.

**Comment by vefimova** [comment]
We need to have a closer look at kubelet's GC (as runtime doesn't suppose to GC by itself, only probably try to resolve conflicts in case of kubelet logic can't deal with it) at https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L360, as I see for dockershim it does its job.
And if it works for us move actual volume creation to StartContainer func as Ivan proposed.

Also re attached volumes -i don't get it, we won't create new domain unless existent one in not stopped.

**Comment by jellonek** [comment]
First issue was already fixed when we started to use tapmanager. I just tested that killing qemu leads to clean start of next one, in which vm has the same network settings working as before start.

**Comment by jellonek** [comment]
Second issue was probably resolved by #556. Now, after calling `poweroff` in vm it's restarted cleanly without any issue.
